smarts/core/road_map.py

```diff
@@ -0,0 +1,437 @@
+            up to lookahead waypoints ahead, starting in this lane at pose.
+            Constrains paths to the supplied route if specified."""
+            raise NotImplementedError()
+
+        def waypoint_paths_at_offset(
+            self, offset: float, lookahead: int = 30, route: RoadMap.Route = None
+        ) -> List[List[Waypoint]]:
+            """Computes equally-spaced Waypoints for all lane paths
+            up to lookahead waypoints ahead, starting offset into this lane.
+            Constrains paths to the supplied route if specified."""
+            raise NotImplementedError()
+
+        @property
+        def features(self) -> List[RoadMap.Feature]:
+            raise NotImplementedError()
+
+        def features_near(self, pose: Pose, radius: float) -> List[RoadMap.Feature]:
+            raise NotImplementedError()
+
+        def buffered_shape(self, width: float = 1.0) -> Polygon:
+            raise NotImplementedError()
+
+        def point_in_lane(self, point: Point) -> bool:
+            raise NotImplementedError()
+
+        def offset_along_lane(self, world_point: Point) -> float:
+            raise NotImplementedError()
+
+        def width_at_offset(self, offset: float) -> float:
+            raise NotImplementedError()
+
+        def from_lane_coord(self, lane_point: RefLinePoint) -> Point:
+            raise NotImplementedError()
+
+        ## The next 6 methods are "reference" implementations for convenience.
+        ## Derived classes may want to extend as well as add a cache.
+
+        def to_lane_coord(self, world_point: Point) -> RefLinePoint:
+            s = self.offset_along_lane(world_point)
+            vector = self.vector_at_offset(s)
+            normal = np.array([-vector[1], vector[0], 0])
+            center_at_s = self.from_lane_coord(RefLinePoint(s=s))
+            offcenter_vector = np.array(world_point) - center_at_s
+            t_sign = np.sign(np.dot(offcenter_vector, normal))
+            t = np.linalg.norm(offcenter_vector) * t_sign
+            return RefLinePoint(s=s, t=t)
+
+        def center_at_point(self, point: Point) -> Point:
+            offset = self.offset_along_lane(point)
+            return self.from_lane_coord(RefLinePoint(s=offset))
+
+        def edges_at_point(self, point: Point) -> Tuple[Point, Point]:
+            offset = self.offset_along_lane(point)
+            width = self.width_at_offset(offset)
+            left_edge = RefLanePoint(s=offset, t=width / 2)
+            right_edge = RefLanePoint(s=offset, t=-width / 2)
+            return (self.from_lane_coord(left_edge), self.from_lane_coord(right_edge))
+
+        def vector_at_offset(self, start_offset: float) -> np.ndarray:
+            add_offset = 1  # a little further down the lane
+            end_offset = start_offset + add_offset
+            # assert end_offset <= self.length + add_offset, f"Offset={end_offset} goes out further than the end of lane=({self.lane_id}, length={length})"
```

Comment:
Actually, this is something we need to look into more.  I commented it out because I hit it during testing and wanted to move past that to look into other things.  But I am not sure my hitting it indicates a real bug or whether this is just an "obsolete" check.  So I need to look into this more before removing or un-commenting it.

Reply:
Ok, I checked and there are cases where I want to allow doing this beyond the end of the lane.  So I removed the assert.